bluez-async/src/modalias.rs

```diff
@@ -0,0 +1,202 @@
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::{self, Display, Formatter};
+use std::str::FromStr;
+use thiserror::Error;
+
+/// An error parsing a [`Modalias`] from a string.
+#[derive(Clone, Debug, Error, Eq, PartialEq)]
+#[error("Error parsing modalias string {0:?}")]
+pub struct ParseModaliasError(String);
+
+/// A parsed modalias string.
+///
+/// For now only the USB subtype is supported.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Modalias {
+    pub vendor_id: u16,
+    pub product_id: u16,
+    pub device_id: u16,
+}
+
+impl Display for Modalias {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "usb:v{:04X}p{:04X}d{:04X}",
+            self.vendor_id, self.product_id, self.device_id
+        )
+    }
+}
+
+impl FromStr for Modalias {
+    type Err = ParseModaliasError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        RawModalias::from_str(s)?
+            .try_into()
+            .map_err(|_| ParseModaliasError(s.to_owned()))
+    }
+}
+
+impl TryFrom<RawModalias> for Modalias {
+    type Error = ();
+
+    fn try_from(raw: RawModalias) -> Result<Self, Self::Error> {
+        if raw.subtype != "usb" {
+            return Err(());
```

Comment:
without knowing anything about modalias is this going to cause problems if someone tries to use a device that isn't subtype=usb?